### PR TITLE
fix(#2449): ActionBar: Restore focus to SelectableCollection on Esc

### LINKED ISSUE #2449: ActionBar: Esc key to close ActionGroupMenu within ActionBar clears selection and closes ActionBar
# 🐛 Bug Report 

At https://github.com/adobe/react-spectrum/blob/2731ebc7d1992b167e3bbab6171e28f2b0adb7fa/packages/%40react-spectrum/actionbar/src/ActionBar.tsx#L71-L78,

The Esc key to close the ActionBar always clears selection. So if the Esc key is triggered from within the ActionGroupMenu to close the Menu of action items that don't fix within the ActionBar, the current selection will be lost, rather than maintaining selection and restoring focus the the ActionGroup menu trigger.

Also, unlike closing the ActionBar by clicking the Dismiss button, pressing the Esc key does not restore focus to the TableView, but instead focus is lost to the `document.body`.

## 🤔 Expected Behavior

Esc key handler should first check that the current focus is actually within the ActionBar, and not within the Portal containing the ActionGroupMenu, before clearing the selection.

When selection is cleared using the Esc key, focus should be restored to the TableView, this may relate to #2444.

## 😯 Current Behavior

The Esc key to close the ActionBar always clears selection. So if the Esc key is triggered from within the ActionGroupMenu to close the Menu of action items that don't fix within the ActionBar, the current selection will be lost, rather than maintaining selection and restoring focus the the ActionGroup menu trigger.

## 💁 Possible Solution

Add a test to make sure the target for the Esc keyboard event is a descendant of the ActionBar before clearing selection.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  @react-spectrum/actionbar v3.0.0-alpha.3
| Browser          | All
| Operating System | All

## 🧢 Your Company/Team

Adobe/Accessibility

### PR DESCRIPTION
**Pull request by majornista**
Closes #2449 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding React Spectrum GitHub Issue #2449. 
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. Open https://reactspectrum.blob.core.windows.net/reactspectrum/973b3aaf0fbb9bd7407c220817dc5b899e8f7952/storybook/index.html?path=/story/actionbar--default
2. Keyboard navigate to an item, or the checkbox for the item, in the TableView.
3. Press Space key to select the item within the TableView, which should expand the ActionBar.
4. Press Tab key to navigate to first action button within the ActionBar toolbar.
5. Press ArrowRight key to move focus to the "More Actions" menu button.
6. Press Space key to expand the "More Actions" menu.
7. Press Esc key to close the "More Actions" menu.
8. Focus should restore to the "More Actions" menu button without clearing the TableView selection.
9. Press Esc key again to clear selections and dismiss the ActionBar.
10. Focus should restore to the item, or the checkbox or cell within the item, in the TableView where focus left off.

It may be useful testing the following behavior using a tool such as [NerdeFocus](https://chrome.google.com/webstore/detail/nerdefocus/lpfiljldhgjecfepfljnbjnbjfhennpd?hl=en-US) to log items as they receive focus.

12. Press Space key to select the item within the TableView again, which should expand the ActionBar.
13. Scroll the selected item out of view within the TableView body.
14. Because the collection is virtualized, the focus will shift the the TableView itself.
15. Press Tab key to navigate to first action button within the ActionBar toolbar.
16. Press ArrowRight key to move focus to the "More Actions" menu button.
17. Press Space key to expand the "More Actions" menu.
18. Press Esc key to close the "More Actions" menu.
19. Focus should restore to the "More Actions" menu button without clearing the TableView selection.
20. Press Esc key again to clear selections and dismiss the ActionBar.
21. Focus should restore to the item, or the checkbox or cell within the item, in the TableView where focus left off, but in order to do so, the TableView will receive focus first, the scroll the appropriate virtualized item into view and then move focus to it.

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx [hunk 1]
```diff
@@ -31,10 +31,27 @@ import {useProviderProps} from '@react-spectrum/provider';
 
 function ActionBar<T extends object>(props: SpectrumActionBarProps<T>, ref: DOMRef<HTMLDivElement>) {
   let isOpen = props.selectedItemCount !== 0;
+  let lastActiveElementRef = useRef(null);
+  let {restoreFocusRef} = props;
```
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx]: I don't think we still need an API review because passing in a `restoreFocusRef` is no longer necessary.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 2]
```diff
@@ -383,7 +383,7 @@ function useAutoFocus(scopeRef: RefObject<HTMLElement[]>, autoFocus: boolean) {
       }
     }
     autoFocusRef.current = false;
-  }, []);
+  }, [scopeRef]);
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/FocusScope.tsx:386]: this change seems unnecessary, what is it for?
what test am i missing? I've added one, for the case you outlined, but it seems like maybe you're covering some others?

Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:386]: > this change seems unnecessary, what is it for?
Fixed an eslint warning for missing dependency on useEffect.
> what test am i missing? I've added one, for the case you outlined, but it seems like maybe you're covering some others?
Is it possible to test the use case for when the focusedKey in the TableView is scrolled out of view?

**Review by ktabors — COMMENTED** [review]
Can we run chromatic on this change?

I also experimented with selecting a couple rows, moving focus to a cell in an unselected row, going into the ActionBar, pressing escape twice and the focus goes to the cell where the last row selection happened, not where focus was within the table.

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 2 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx]: > Is it possible to test the use case for when the focusedKey in the TableView is scrolled out of view?

should be, you can find examples of how to scroll TableView in tests here https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/table/test/Table.test.js#L1478

Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx]: ok, if it's just fixing eslint, i'm fine with it

**Comment by snowystinger** [comment]
@ktabors I think the behavior you're seeing makes sense, if you leave a collection that has selections, even if you looked at a few more items, when you come back, you'll be at either the first or last selected item (depending if you arrived via Tab vs shift+Tab). So even though you cleared the selection with Esc, I think returning to where you had a selection makes sense

File: packages/@react-types/actionbar/src/index.d.ts [hunk 3]
```diff
@@ -10,14 +10,15 @@
...
-import {Key, ReactNode} from 'react';
+import {CollectionBase, DOMProps, DOMRefValue, StyleProps} from '@react-types/shared';
+import {Key, ReactNode, RefObject} from 'react';
 
 export interface ActionBarProps<T> extends CollectionBase<T> {
   // Required. When zero, the ActionBar is hidden.
   selectedItemCount: number | 'all',
   // Also required. When clicked, the selection should be cleared.
   onClearSelection: () => void,
+  restoreFocusRef: RefObject<DOMRefValue<HTMLElement>> | RefObject<HTMLElement>,
```
Comment (snowystinger) [inline packages/@react-types/actionbar/src/index.d.ts:21]: API change I'm not sure about

**Review by snowystinger — DISMISSED** [review]
Approving because it's working in testing and code seems ok, though will require @devongovett / @dannify to check the API change noted in the review

File: packages/@react-types/actionbar/src/index.d.ts (hunk 3 above)
Comment (LFDanLu) [reply packages/@react-types/actionbar/src/index.d.ts:21]: IMO we should remove this and relocate it into the `ActionBarInnerProps` since it is only used for ActionBarInner and shouldn't be visible to external users

**Review by ktabors — COMMENTED** [review]
Since Dob said the behavior I noticed is okay I'm okay approving, but waiting to see about the API change.

File: packages/@react-types/actionbar/src/index.d.ts (hunk 3 above)
Comment (snowystinger) [reply packages/@react-types/actionbar/src/index.d.ts:21]: o, good catch, didn't notice that due to the code folding that github does, that is what we should do

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx [hunk 4]
```diff
@@ -82,6 +82,33 @@ const ActionBarInner = React.forwardRef((props: ActionBarInnerProps, ref: DOMRef
...
   }, [formatMessage]);
 
+  let restoreFocusRef = useRef<HTMLElement>(null);
+
+  // When selectedItemCount changes, update the restoreFocusRef for match the last document.activeElement
+  useLayoutEffect(() => {
+    if (selectedItemCount !== 0 && lastCount.current) {
+      restoreFocusRef.current = document.activeElement as HTMLElement;
+    }
+  }, [selectedItemCount]);
```
Comment (LFDanLu) [inline packages/@react-spectrum/actionbar/src/ActionBar.tsx:92]: Hmm, so we only update the `restoreFocusRef` when an item selection is made? Ideally IMO it would be best if we could track what was the last focused cell/row in the TableView since the user could shift their focus via arrow keys before tabbing into the ActionBar. 

Been playing around with this locally to see if there is a way to do this. One DOM crawly way perhaps would be to find what element has `tabIndex = 0` in ActionBar container (excluding the ActionBar itself) and manually focus that when the ActionBar closes. The downside is that it is makes an assumption that the component wrapped in the ActionBarContainer is a TableView or at least follows the roving tab index pattern (also is quite DOM crawly which we try to avoid). Haven't been able to figure out a different way unfortunately

Comment (LFDanLu) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:92]: Actually, could we not get rid of the the useLayoutEffect and the useEffect below and just do
```
  useEffect(() => {
    return () => {
      let walker = getFocusableTreeWalker(domRef.current.parentElement, {tabbable: true});
      let node = walker.nextNode() as HTMLElement;
      requestAnimationFrame(() => {
        node && node.focus()
      });
    }
  }, [domRef]);
```
Or are we anticipating situations where we don't have a SelectableCollection component wrapped within the ActionBar? Should we then instead record what the last document.activeElement was when focus enters the ActionBar and restore focus to that element if it still exist when the ActionBar unmounts?

Comment (majornista) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:93]: I think I have a solution for this with https://github.com/adobe/react-spectrum/pull/2451/commits/eab13f0bf66288444bd59d01045a3b048d9f1772

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx [hunk 5]
```diff
@@ -82,6 +82,53 @@ const ActionBarInner = React.forwardRef((props: ActionBarInnerProps, ref: DOMRef
...
+  let containerRef = useRef<HTMLElement>(null);
+
+  useEffect(() => {
+    // Event handler to keep track of focus changes within ActionBarContainer when ActionBar is open.
+    let onFocusContainer = (e:FocusEvent) => {
+      let activeElement = e.target as HTMLElement;
+      if (domRef.current && !domRef.current.contains(activeElement)) {
+        restoreFocusRef.current = activeElement;
+      }
+    };
```
Comment (LFDanLu) [inline packages/@react-spectrum/actionbar/src/ActionBar.tsx:95]: This is very close. One last thing I noticed was if the user tabs into the ActionBar and then scrolls the focused key out of view in such a way that it is removed from the DOM, closing the ActionBar doesn't restore focus to the previously focused cell. This is because the restoreFocusRef (i.e. the focused cell) is no longer in the DOM and onFocusContainer doesn't update the restoreFocusRef because the active element is one of the buttons in the ActionBar when the focused key is scrolled out of view.

Perhaps we should bring back the fallback treewalker code when `document.body.contains(restoreFocusRef.current)` returns false in the useEffect cleanup?
```
        let walker = getFocusableTreeWalker(container, {tabbable: true});
        let node = walker.nextNode() as HTMLElement;
        node && node.focus();
```

Repro steps:
1. Select the first row of the ActionBar's TableView via arrow keys + Enter
2. Tab into the ActionBar
3. Scroll to the last row of the TableView via touchpad scroll/scroll wheel
4. Close the ActionBar via escape key or focusing the 'x' and hitting enter. Note focus isn't restored

Comment (snowystinger) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:95]: Unsure if it's the same reason, but you can also use the default story
1. navigate to the 10th item, select it
2. navigate to the 5th item, select it
3. tab to the actionbar
4. press Esc

for me, focus went to item 12, though I think it should have gone to either 5? given the discussion?

File: packages/@react-spectrum/actionbar/test/ActionBar.test.js [hunk 6]
```diff
@@ -134,4 +135,92 @@ describe('ActionBar', () => {
...
+
+    expect(document.activeElement).toBe(menu);
+
+    fireEvent.keyDown(document.activeElement, {key: 'Escape'});
+    fireEvent.keyUp(document.activeElement, {key: 'Escape'});
+
+    expect(document.activeElement).toBe(moreButton);
+
+    act(() => {
+      fireEvent.keyDown(document.activeElement, {key: 'Escape'});
```
Comment (snowystinger) [inline packages/@react-spectrum/actionbar/test/ActionBar.test.js:212]: fire event is already wrapped in an act, you should only need to wrap the runAllTimers

**Comment by majornista** [comment]
Having some difficulty with ci test `test-17` failing: https://app.circleci.com/pipelines/github/adobe/react-spectrum/6313/workflows/73bfdf64-a59b-4f56-8e13-8eacd46c83bf/jobs/55380, I skipped it. The test passes locally.

**Comment by snowystinger** [comment]
I'm still able to easily get focused on the wrong element following the steps outlined here https://github.com/adobe/react-spectrum/pull/2451#discussion_r749610381
Not only that, but I end up on different rows from run to run

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx (hunk 5 above)
Comment (LFDanLu) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:95]: Hm I'm having a hard time reproducing in https://reactspectrum.blob.core.windows.net/reactspectrum/33780525ea16cfe5c1c0a87dfba2b8cf855901b0/storybook/index.html?path=/story/actionbar--default. Can you still reproduce?

**Comment by LFDanLu** [comment]
@majornista Yep that is the test suite being run in react 17, I can reproduce it locally after doing `yarn install-17` and running the tests again, trying to dig into why it is failing.

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx (hunk 5 above)
Comment (LFDanLu) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:95]: Ah nvm, I'm seeing it in the react 17 storybook, hmmm

Comment (LFDanLu) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:95]: Looks like the issue is when the user tabs out of the TableView and into the ActionBar, `useSelectableCollection` will [focus the last tabbable item](https://github.com/adobe/react-spectrum/blob/b34e500ddd6ca860cbc943f90a85ea911f2d881b/packages/%40react-aria/selection/src/useSelectableCollection.ts#L237-L266) so that we can exit the table as a single tab stop. This gets picked up by `onFocusContainer` and is set as the element to restore focus to hence the behavior observed by @snowystinger.

Comment (LFDanLu) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:95]: @majornista Ah scratch that, perhaps the fix is just to remove the `restoreFocus` prop from ActionBar's FocusScope. My previous comment is a red herring I think since `useSelectableItem` will properly move the focus back to the tracked focusedKey I believe via its useEffect. However, FocusScope's `restoreFocus` is problematic here because it tries to restore focus back to the item that was focused when the ActionBar first mounted but that isn't necessary the right place to restore focus to since the user can move focus before tabbing into the ActionBar. Since we are basically handling restoring focus via the ActionBar useEffect logic we can do away with the ActionBar's FocusScope `restoreFocus` IMO.

**Comment by LFDanLu** [comment]
@majornista Hrm, looks like the skipped test fails in React 17 because the activeElement is already on the proper row after the 2nd to last `runAllTimers`. Making the requestAnimationFrame mock properly async via `jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => setTimeout(cb, 0));` in the test and then removing 
```
    expect(document.activeElement).toBe(table);
    fireEvent.focus(table);

    act(() => jest.runAllTimers());
```
from the latter part of the test makes it pass for both React 16 and 17 but I would need to dig some more into why exactly that is. Perhaps a similar issue happened to this test that prompted the need for the same mocking of requestAnimationFrame https://github.com/adobe/react-spectrum/blob/2e8faba3c880b8ffcdde4790207cc65db8522c7a/packages/%40react-spectrum/table/test/Table.test.js#L2938-L2939

**Comment by majornista** [comment]
> @majornista Hrm, looks like the skipped test fails in React 17 because the activeElement is already on the proper row after the 2nd to last `runAllTimers`. Making the requestAnimationFrame mock properly async via `jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => setTimeout(cb, 0));` in the test and then removing
> 
> ```
>     expect(document.activeElement).toBe(table);
>     fireEvent.focus(table);
> 
>     act(() => jest.runAllTimers());
> ```
> 
> from the latter part of the test makes it pass for both React 16 and 17 but I would need to dig some more into why exactly that is. Perhaps a similar issue happened to this test that prompted the need for the same mocking of requestAnimationFrame
> 
> https://github.com/adobe/react-spectrum/blob/2e8faba3c880b8ffcdde4790207cc65db8522c7a/packages/%40react-spectrum/table/test/Table.test.js#L2938-L2939

Interesting, but I also just noticed that I can reproduce the behaviors @snowystinger mentions in https://github.com/adobe/react-spectrum/pull/2451#issuecomment-969436922 on [storybook-17](https://reactspectrum.blob.core.windows.net/reactspectrum/33780525ea16cfe5c1c0a87dfba2b8cf855901b0/storybook-17/index.html?path=/story/actionbar--default), but not on [storybook](https://reactspectrum.blob.core.windows.net/reactspectrum/33780525ea16cfe5c1c0a87dfba2b8cf855901b0/storybook/index.html?path=/story/actionbar--default), so there is an issue.

**Comment by LFDanLu** [comment]
@majornista Yep, the broken behavior on the react 17 storybook might be fixed via https://github.com/adobe/react-spectrum/pull/2451#discussion_r750694193 (at least it seems to work on my local testing when I removed restoreFocus)

**Comment by majornista** [comment]
@LFDanLu  and @snowystinger I think I may have a fix for the React 17 issues: https://github.com/adobe/react-spectrum/pull/2451/commits/72a465e4cf25ce92deb638bacbe7323497e11757

File: packages/@react-spectrum/actionbar/src/ActionBar.tsx [hunk 7]
```diff
@@ -82,6 +82,67 @@ const ActionBarInner = React.forwardRef((props: ActionBarInnerProps, ref: DOMRef
...
+      if (timeoutId) {
+        cancelAnimationFrame(timeoutId);
+        timeoutId = null;
+      }
+      timeoutId = requestAnimationFrame(() => {
+        if (domRef.current && !domRef.current.contains(activeElement)) {
+          restoreFocusRef.current = activeElement;
+        }
+      });
+    };
```
Comment (LFDanLu) [inline packages/@react-spectrum/actionbar/src/ActionBar.tsx:102]: Still seeing issues with where focus is being restored.
1. Using the keyboard select row 1
2. Move focus to "Foo 3"
3. Tab to ActionBar
4. Scroll to bottom of TableView via scroll wheel/touch pad
5. Hit Escape
Focus moves to row 13 or some other rows. Did https://github.com/adobe/react-spectrum/pull/2451#discussion_r750694193 not work when you tried it?

Comment (majornista) [reply packages/@react-spectrum/actionbar/src/ActionBar.tsx:102]: Interesting. Looks like adding the cancelAnimationFrame works for [Storybook-17](https://reactspectrum.blob.core.windows.net/reactspectrum/72a465e4cf25ce92deb638bacbe7323497e11757/storybook-17/index.html?path=/story/actionbar--default), but breaks [Storybook](https://reactspectrum.blob.core.windows.net/reactspectrum/72a465e4cf25ce92deb638bacbe7323497e11757/storybook/index.html?path=/story/actionbar--default) now.

File: packages/@react-aria/overlays/src/useOverlay.ts [hunk 8]
```diff
@@ -111,6 +111,7 @@ export function useOverlay(props: OverlayProps, ref: RefObject<HTMLElement>): Ov
   // Handle the escape key
   let onKeyDown = (e) => {
     if (e.key === 'Escape' && !isKeyboardDismissDisabled) {
+      e.stopPropagation();
```
Comment (snowystinger) [inline packages/@react-aria/overlays/src/useOverlay.ts:114]: this seems separate from the rest, less controversial, and higher priority. should this just be its own PR to get in easily?

Comment (majornista) [reply packages/@react-aria/overlays/src/useOverlay.ts:114]: See https://github.com/adobe/react-spectrum/pull/2976
